src/opencl_sappse_fmt_plug.c

```diff
@@ -0,0 +1,311 @@
+
+		// Initialize openCL tuning (library) for this format.
+		opencl_init_auto_setup(SEED, 0, NULL, warn, 1, self,
+		                       create_clobj, release_clobj,
+		                       sizeof(sappse_password), 0, db);
+
+		// Auto tune execution from shared/included code.
+		autotune_run(self, 1, 0, 200);
+	}
+}
+
+static void done(void)
+{
+	if (autotuned) {
+		release_clobj();
+
+		HANDLE_CLERROR(clReleaseKernel(crypt_kernel), "Release kernel");
+		HANDLE_CLERROR(clReleaseProgram(program[gpu_id]), "Release Program");
+
+		autotuned--;
+	}
+}
+
+static void set_salt(void *salt)
+{
+	cur_salt = (struct custom_salt*)salt;
+
+	currentsalt.salt_size = cur_salt->salt_size;
+	currentsalt.iterations = cur_salt->iterations;
+	currentsalt.encrypted_pin_size = cur_salt->encrypted_pin_size;
+	memcpy((char*)currentsalt.salt, cur_salt->salt, currentsalt.salt_size);
+	memcpy((char*)currentsalt.encrypted_pin, cur_salt->encrypted_pin, currentsalt.encrypted_pin_size);
+
+	HANDLE_CLERROR(clEnqueueWriteBuffer(queue[gpu_id], mem_setting,
+		CL_FALSE, 0, settingsize, &currentsalt, 0, NULL, NULL),
+	    "Copy setting to gpu");
+}
+
+static void sappse_set_key(char *key, int index)
+{
+	uint32_t length = strlen(key);
+
+	if (length > PLAINTEXT_LENGTH)
+		length = PLAINTEXT_LENGTH;
+	inbuffer[index].length = length;
+	memcpy(inbuffer[index].v, key, length);
+}
+
+static char *get_key(int index)
+{
+	static char ret[PLAINTEXT_LENGTH + 1];
+	uint32_t length = inbuffer[index].length;
+
+	memcpy(ret, inbuffer[index].v, length);
+	ret[length] = '\0';
+
+	return ret;
+}
+
+static int crypt_all(int *pcount, struct db_salt *salt)
+{
+	const int count = *pcount;
+	size_t gws = count;
+	size_t *lws = (local_work_size && !(gws % local_work_size)) ?
+		&local_work_size : NULL;
+
+	// Copy data to gpu
+	BENCH_CLERROR(clEnqueueWriteBuffer(queue[gpu_id], mem_in, CL_FALSE, 0,
+		insize, inbuffer, 0, NULL, multi_profilingEvent[0]),
+		"Copy data to gpu");
+
+	// Run kernel
+	BENCH_CLERROR(clEnqueueNDRangeKernel(queue[gpu_id], crypt_kernel, 1,
+		NULL, &gws, lws, 0, NULL,
+		multi_profilingEvent[1]),
+		"Run kernel");
+
+	// Read the result back
+	BENCH_CLERROR(clEnqueueReadBuffer(queue[gpu_id], mem_out, CL_TRUE, 0, outsize, output, 0, NULL, multi_profilingEvent[5]), "Copy result back");
+
+	return count;
+}
+
+static int cmp_all(void *binary, int count)
+{
+	int index;
+
+	for (index = 0; index < count; index++)
+		if (output[index].cracked)
+			return 1;
+	return 0;
+}
+
+static int cmp_one(void *binary, int index)
+{
+	return output[index].cracked;
+}
+
+static int cmp_exact(char *source, int index)
+{
+	return 1;
+}
+
+struct fmt_main fmt_opencl_sappse = {
+	{
+		FORMAT_LABEL,
+		FORMAT_NAME,
+		ALGORITHM_NAME,
+		BENCHMARK_COMMENT,
+		BENCHMARK_LENGTH,
+		0,
+		PLAINTEXT_LENGTH,
+		BINARY_SIZE,
+		BINARY_ALIGN,
+		SALT_SIZE,
+		SALT_ALIGN,
+		MIN_KEYS_PER_CRYPT,
+		MAX_KEYS_PER_CRYPT,
+		FMT_CASE | FMT_8_BIT | FMT_HUGE_INPUT,
```

Comment:
You really need to get the flags right.

Perhaps you should write a checklist for what to always look out for after copying a template format (and then actually use it)? And anyway you should review your own code, just as if you reviewed someone else's code. I usually do so myself before committing.